### PR TITLE
Add new error types for document additions

### LINKED ISSUE #1690: Handle NDJSON format
Sub issue of https://github.com/meilisearch/MeiliSearch/issues/1593

⚠️ All the implementation details are available in this [section of the specification](https://github.com/meilisearch/specifications/blob/develop/text/0029-indexing-ndjson.md). Please take a look before starting the implementation.

---

## Basic behavior 

When adding or updating documents (`POST/PUT /indexes/:uid/documents`), MeiliSearch should accept the `Content-Type: application/x-ndjson` and the corresponding data should be sent in NDJSON format.
Cf [this part in the spec](https://github.com/meilisearch/specifications/blob/develop/text/0029-indexing-ndjson.md#api-endpoints)

## Error handler

Also related to [this issue](https://github.com/meilisearch/MeiliSearch/issues/1665)

⚠️ All the errors are sent synchronously

List of expected errors related to the behavior:
  - Omitted `Content-Type` header will lead to a `415 Unsupported Media Type` - `missing_content_type` error code.
  - Sending an empty `Content-Type` will lead to a `415 Unsupported Media Type` - `invalid_content_type error` code.
  - Sending a different `Content-Type` than `application/json`, `application/x-ndjson` or `text/csv` will lead to `415 Unsupported Media Type` -  `invalid_content_type` error code.
  - Sending an empty payload will lead to a `400 Bad Request` - `missing_payload` error code.
  - Sending a different payload type than the `Content-Type` header should return a 400 Bad Request - `malformed_payload` error code.
  - Sending a payload excessing the limit will lead to a `413 Payload Too Large` - `payload_too_large` error code.
  - Sending an invalid ndjson format will lead to a `400 bad_request` - `malformed_payload` error code. 

The expected implementation of the different error:

### `missing_content_type`

HTTP Code: `415 Unsupported Media Type`

```json
{
    "message": "A Content-Type header is missing. Accepted values for the Content-Type header are: :contentTypeList",
    "code": "missing_content_type",
    "type": "invalid_request",
    "link": "https://docs.meilisearch.com/errors#missing_content_type"
}
```

### `invalid_content_type`

HTTP Code: `415 Unsupported Media Type`

```json
{
    "message": "The Content-Type :contentType is invalid. Accepted values for the Content-Type header are: :contentTypeList",
    "code": "invalid_content_type",
    "type": "invalid_request",
    "link": "https://docs.meilisearch.com/errors#invalid_content_type"
}
```

### `missing_payload`

HTTP Code: `400 Bad Request`

```json
{
    "message": "A :payloadType payload is missing.",
    "code": "missing_payload",
    "type": "invalid_request",
    "link": "https://docs.meilisearch.com/errors#missing_payload"
}
```

### `malformed_payload`

HTTP Code: `400 Bad Request`

```json
"message": ":syntaxErrorHelper. The :payloadType payload provided is malformed.",
"code": "malformed_payload",
"type": "invalid_request",
"link": "https://docs.meilisearch.com/errors#malformed_payload"
```

### LINKED ISSUE #1691: Handle CSV format
Sub issue of https://github.com/meilisearch/MeiliSearch/issues/1593

⚠️ All the implementation details are available in this [section of the specification](https://github.com/meilisearch/specifications/blob/develop/text/0028-indexing-csv.md). Please take a look before starting the implementation.

---

## Basic behavior 

When adding or updating documents (`POST/PUT /indexes/:uid/documents`), MeiliSearch should accept the `Content-Type: text/csv` and the corresponding data should be sent in CSV format.
Cf [this part in the spec](https://github.com/meilisearch/specifications/blob/develop/text/0028-indexing-csv.md#api-endpoints)

The formatting rules for CSV accepted by MeiliSearch are described [here in the spec](https://github.com/meilisearch/specifications/blob/develop/text/0028-indexing-csv.md#csv-formatting-rules).

## Error handler

Also related to [this issue](https://github.com/meilisearch/MeiliSearch/issues/1665)

⚠️ All the errors are sent synchronously

List of expected errors related to the behavior:
  - Omitted `Content-Type` header will lead to a `415 Unsupported Media Type` - `missing_content_type` error code.
  - Sending an empty `Content-Type` will lead to a `415 Unsupported Media Type` - `invalid_content_type error` code.
  - Sending a different `Content-Type` than `application/json`, `application/x-ndjson` or `text/csv` will lead to `415 Unsupported Media Type` -  `invalid_content_type` error code.
  - Sending an empty payload will lead to a `400 Bad Request` - `missing_payload` error code.
  - Sending a different payload type than the `Content-Type` header should return a 400 Bad Request - `malformed_payload` error code.
  - Sending a payload excessing the limit will lead to a `413 Payload Too Large` - `payload_too_large` error code.
  - Sending an invalid CSV format will lead to a `400 bad_request` - `malformed_payload` error code. 

The expected implementation of the different error:

### `missing_content_type`

HTTP Code: `415 Unsupported Media Type`

```json
{
    "message": "A Content-Type header is missing. Accepted values for the Content-Type header are: :contentTypeList",
    "code": "missing_content_type",
    "type": "invalid_request",
    "link": "https://docs.meilisearch.com/errors#missing_content_type"
}
```

### `invalid_content_type`

HTTP Code: `415 Unsupported Media Type`

```json
{
    "message": "The Content-Type :contentType is invalid. Accepted values for the Content-Type header are: :contentTypeList",
    "code": "invalid_content_type",
    "type": "invalid_request",
    "link": "https://docs.meilisearch.com/errors#invalid_content_type"
}
```

### `missing_payload`

HTTP Code: `400 Bad Request`

```json
{
    "message": "A :payloadType payload is missing.",
    "code": "missing_payload",
    "type": "invalid_request",
    "link": "https://docs.meilisearch.com/errors#missing_payload"
}
```

### `malformed_payload`

HTTP Code: `400 Bad Request`

```json
"message": ":syntaxErrorHelper. The :payloadType payload provided is malformed.",
"code": "malformed_payload",
"type": "invalid_request",
"link": "https://docs.meilisearch.com/errors#malformed_payload"
```

### PR DESCRIPTION
**Pull request by MarinPostma**
Adds the missing errors for the documents routes, as specified.

close #1691
close #1690

### CODE REVIEW

File: meilisearch-http/src/error.rs [hunk 1]
```diff
@@ -9,6 +9,23 @@ use aweb::error::{JsonPayloadError, QueryPayloadError};
 use meilisearch_error::{Code, ErrorCode};
 use serde::{Deserialize, Serialize};
 
+#[derive(Debug, thiserror::Error)]
+pub enum MeilisearchHttpError {
+    #[error("A Content-Type header is missing. Accepted values for the Content-Type header are: \"application/json\", \"application/x-ndjson\", \"test/csv\"")]
+    MissingContentType,
+    #[error("The Content-Type \"{0}\" is invalid. Accepted values for the Content-Type header are: \"application/json\", \"application/x-ndjson\", \"test/csv\"")]
```
Comment (Kerollmops) [inline meilisearch-http/src/error.rs:16]:
```suggestion
    #[error("A Content-Type header is missing. Accepted values for the Content-Type header are: \"application/json\", \"application/x-ndjson\", \"text/csv\"")]
    MissingContentType,
    #[error("The Content-Type \"{0}\" is invalid. Accepted values for the Content-Type header are: \"application/json\", \"application/x-ndjson\", \"text/csv\"")]
```

File: meilisearch-http/src/routes/indexes/documents.rs [hunk 2]
```diff
@@ -165,127 +124,76 @@ pub struct UpdateDocumentsQuery {
...
-
-pub async fn update_documents_ndjson(
+pub async fn update_documents(
     meilisearch: GuardedData<Private, MeiliSearch>,
     path: web::Path<IndexParam>,
     params: web::Query<UpdateDocumentsQuery>,
     body: Payload,
+    req: HttpRequest,
 ) -> Result<HttpResponse, ResponseError> {
+    debug!("called with params: {:?}", params);
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/indexes/documents.rs:155]: Do you want to keep that?

**Review by Kerollmops — CHANGES_REQUESTED** [review]
No sure about what we want to do but an empty payload is valid when it is an `ndjson` payload.

File: meilisearch-http/src/routes/indexes/documents.rs (hunk 2 above)
Comment (MarinPostma) [reply meilisearch-http/src/routes/indexes/documents.rs:155]: Yes, this yes, this is part of every route I think. @Irevoire made a PR some time ago to add them

Comment (irevoire) [reply meilisearch-http/src/routes/indexes/documents.rs:155]: Yes, this is in the spec about the logs.

**Comment by MarinPostma** [comment]
> No sure about what we want to do but an empty payload is valid when it is an ndjson payload.

@Kerollmops  This is part of the specification, and I have no way to distinguish between an empty payload and no payload at all...

**Review by Kerollmops — APPROVED** [review]

**Review by curquiza — CHANGES_REQUESTED** [review]
Only the POST and PUT `/documents` routes should accept CSV end NDJSON. Other routes with payload should only accept `application/json`

When I do
```
curl -i \
               -H 'Content-type: application/x-ndjson' \
               -X POST 'http://localhost:7700/indexes/indexUID/settings' \
               --data-binary '{ "filterableAttributes": ["title"] }'
```

I should get an error `unsupported_media_type`. Only `application/json` is accepted for this kind of route.

Also

```
curl -i \
               -H 'Content-type: application/lol' \
               -X POST 'http://localhost:7700/indexes/indexUID/settings' \
               --data-binary '{ "filterableAttributes": ["title"] }'
```

Should return a `unsupported_media_type` error as well

**Comment by curquiza** [comment]
Or maybe this will be in another PR? More retated to this issue https://github.com/meilisearch/MeiliSearch/issues/1665?

**Comment by MarinPostma** [comment]
> Or maybe this will be fixed in another PR more related to this issue #1665?

@curquiza yes, this is another issue

**Review by irevoire — APPROVED** [review]
lgtm

**Review by curquiza — APPROVED** [review]
Good to me then!

bors merge 🚀

**Comment by irevoire** [comment]
bors merge
and learn how to rebase

**Review by curquiza — APPROVED** [review]
bors merge

**Pull request merged**